File: scripts/fetch_benchmarks.py
#!/usr/bin/env python3
"""获取美股ETF基准数据 (在GitHub Actions美国服务器上跑)"""
import json, os, sys
from datetime import datetime

# 用 yfinance 或者手动请求
try:
    import yfinance as yf
    HAS_YF = True
except:
    HAS_YF = False

BENCHES = {
    'qqq': {'name': '纳斯达克100(QQQ)', 'symbol': 'QQQ'},
    'spy': {'name': '标普500(SPY)', 'symbol': 'SPY'},
    'qqqi': {'name': 'QQQI', 'symbol': 'QQQI'},
}

OUT = os.path.join(os.path.dirname(__file__), '..', 'benchmarks')

def fetch_yfinance(symbol):
    """用 yfinance 获取历史数据"""
    t = yf.Ticker(symbol)
    hist = t.history(period='5y')
    data = []
    for idx, row in hist.iterrows():
        data.append({
            'date': idx.strftime('%Y-%m-%d'),
            'price': round(float(row['Close']), 2),
        })
    return data

def fetch_fallback(symbol):
    """不用yfinance的备选方案"""
    import urllib.request
    # 用 Yahoo Finance CSV endpoint
    import time
    ts = int(time.time())
    url = f'https://query1.finance.yahoo.com/v7/finance/download/{symbol}?period1=1577836800&period2={ts}&interval=1d&events=history'
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urllib.request.urlopen(req, timeout=15) as r:
        lines = r.read().decode().strip().split('\n')
    data = []
    for line in lines[1:]:  # skip header
        parts = line.split(',')
        if len(parts) >= 5 and parts[4] and parts[4] != 'null':
            data.append({'date': parts[0], 'price': round(float(parts[4]), 2)})
    return data

def main():
    if not os.path.exists(OUT):
        os.makedirs(OUT)
    
    for key, cfg in BENCHES.items():
        path = os.path.join(OUT, f'{key}.json')
        print(f'{cfg["name"]}...', end=' ')
        
        try:
            if HAS_YF:
                data = fetch_yfinance(cfg['symbol'])
            else:
                data = fetch_fallback(cfg['symbol'])
            
            if len(data) > 100:
                out = {
                    'name': cfg['name'],
                    'symbol': key,
                    'updated': datetime.now().isoformat(),
                    'data': data,
                }
                with open(path, 'w', encoding='utf-8') as f:
                    json.dump(out, f, ensure_ascii=False)
                print(f'✅ {len(data)} 条')
            else:
                print(f'❌ 数据不足 ({len(data)})')
        except Exception as e:
            print(f'❌ {e}')

if __name__ == '__main__':
    main()

File: scripts/fetch_benchmarks.js
// 获取基准数据: QQQ, SPY, CSI300, CSI500
// 运行: node scripts/fetch_benchmarks.js
// 输出到 benchmarks/ 目录

const https = require('https');
const fs = require('fs');
const path = require('path');

const BENCHES = {
    'qqq': { yahoo: 'QQQ', name: '纳斯达克100', div: 0.0065 },
    'spy': { yahoo: 'SPY', name: '标普500', div: 0.013 },
    'qqqi': { yahoo: 'QQQI', name: 'QQQI', div: 0.12 },
    'csi300': { sina: 'sh000300', name: '沪深300', div: 0.02 },
    'csi500': { sina: 'sh000905', name: '中证500', div: 0.015 },
};

const OUT_DIR = path.join(__dirname, '..', 'benchmarks');

function fetch(url) {
    return new Promise((resolve, reject) => {
        const opts = { headers: { 'User-Agent': 'Mozilla/5.0' }, timeout: 15000 };
        https.get(url, opts, res => {
            let data = '';
            res.on('data', c => data += c);
            res.on('end', () => resolve(data));
        }).on('error', reject);
    });
}

async function fetchYahoo(symbol) {
    const url = `https://query1.finance.yahoo.com/v8/finance/chart/${symbol}?range=5y&interval=1d`;
    const raw = await fetch(url);
    const json = JSON.parse(raw);
    const result = json.chart?.result?.[0];
    if (!result) throw new Error('yahoo empty');
    
    const timestamps = result.timestamp;
    const quotes = result.indicators.quote?.[0];
    if (!quotes) throw new Error('no quotes');
    
    const data = [];
    for (let i = 0; i < timestamps.length; i++) {
        if (quotes.close?.[i]) {
            data.push({
                date: new Date(timestamps[i] * 1000).toISOString().slice(0, 10),
                price: quotes.close[i],
            });
        }
    }
    return data;
}

async function fetchSina(code) {
    const url = `https://web.ifzq.gtimg.cn/appstock/app/fqkline/get?param=${code},day,,,800,qfq`;
    const raw = await fetch(url);
    const json = JSON.parse(raw);
    const key = code.startsWith('sh') ? 'sh' : 'sz';
    const days = json?.data?.[code]?.day || json?.data?.[code]?.qfqday || [];
    
    return days.map(d => ({
        date: d[0].replace(/-/g, '-'),
        price: parseFloat(d[2]), // 收盘价
    })).filter(d => d.price > 0);
}

async function main() {
    if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR, { recursive: true });
    
    for (const [key, cfg] of Object.entries(BENCHES)) {
        console.log(`获取 ${cfg.name}...`);
        let data = [];
        try {
            if (cfg.yahoo) data = await fetchYahoo(cfg.yahoo);
            else if (cfg.sina) data = await fetchSina(cfg.sina);
            
            if (data.length > 50) {
                // 保存数据
                const out = { name: cfg.name, symbol: key, divYield: cfg.div, updated: new Date().toISOString(), data };
                fs.writeFileSync(path.join(OUT_DIR, `${key}.json`), JSON.stringify(out, null, 2));
                console.log(`  ✅ ${data.length} 条`);
            } else {
                console.log(`  ❌ 数据不足 (${data.length})`);
            }
        } catch (e) {
            console.log(`  ❌ ${e.message}`);
        }
    }
}

main().catch(console.error);
